### ISSUE TITLE
Adding javascriptreact / typescriptreact filetypes

### ISSUE THREAD

**Issue by ecklf** [vim/vim#4830]
**Is your feature request related something that is currently hard to do? Please describe.**
I assume all plugins trying to fully support Microsofts language server protocol are affected by this. As you can see in the [LSP specification](https://microsoft.github.io/language-server-protocol/specification) there are filetypes named `javascriptreact` and `typescriptreact`. This is also the case for [VSCode](https://code.visualstudio.com/docs/languages/identifiers) since it is built upon that. 

Taking a look at my autocomplete plugin of choice (coc.nvim), they also seem to require to set either of those filetypes f.ex [(javascriptreact or javascript.jsx)](https://github.com/neoclide/coc.nvim/blob/b9967440b806a20067342b2ecc09682e6e1c685c/src/model/document.ts#L86) in order to have support for jsx/tsx specific snippets. This is also confirmed by a maintainer in the coc-tsserver [readme](https://github.com/neoclide/coc-tsserver).

**Describe the solution you'd like**
Automatically setting jsx / tsx files to:
`javascriptreact / typescriptreact`
or 
`javascript.jsx / typescript.tsx`

**Describe alternatives you've considered**
Setting filetypes manually
```
autocmd bufnewfile,bufread *.tsx set filetype=typescript.tsx
autocmd bufnewfile,bufread *.jsx set filetype=javascript.jsx
```

Would there be any drawbacks on making those filetypes official? 
I would love to hear your opinions on this.

Edit* Removing instructions line

**Comment by brammool** [comment]
Would there be different syntax or indent files for typescriptreact and javascriptreact?

**Comment by ecklf** [comment]
@brammool jsx is a syntax extension for javascript, so there is additional syntax and indenting. 

There are plugins like [vim-jsx-pretty](https://github.com/MaxMEllon/vim-jsx-pretty) to support this, however they require you to set the filetype to `javascript.jsx` to function. The initial idea for this request was to add jsx specific comment syntax support (within your xml/html markup) for [nerdcommenter](https://github.com/scrooloose/nerdcommenter/pull/393)

Edit* fix typo

**Comment by brammool** [comment]
> @brammool jsx is a syntax extension for javascript, so there is additional syntax and indenting. 
> 
> There are plugins like
> [vim-jsx-pretty](https://github.com/MaxMEllon/vim-jsx-pretty) to
> support this, however they require you to set the filetype to
> `javascript.jsx` to function. The initial idea for this request was to
> adding support jsx specific comment syntax (within your xml/html
> markup) for
> [nerdcommenter](https://github.com/scrooloose/nerdcommenter/pull/393)

The dot notation for a filetype/syntax name is rarely used and most
users won't know how it works.  I would prefer using a normal name,
unless the name with a dot has a clear advantage.

The example uses ".doxygen", because doxygen is something that can be
added on top of multiple other syntaxes.  Thus instead of making several
"doxygen aware" syntaxes, it's added on top.  That works, because
doxygen doesn't change the syntax, it only adds more items (comment
stuff in this case).  When the base syntax changes, doxygen doesn't need
to be adjusted, it keeps adding the same things.

So with .jsx and .tsx files, is this adding something generic on top of
javascript and typescript, or is it changing syntax items?  Adding
".jsx" and ".tsx" suggests the additions are specific for javascript and
typescript, in which case it's better to use "javascriptreact" and
"typescriptreact", which then wrap the base syntax and add their own
"react specific" stuff.  That is also more flexible.  But it may require
some more lines of Vim script.

-- 
How many light bulbs does it take to change a person?

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by alerque** [comment]
@brammool I believe the React language changes are "additive" in the sense that they start with and extend the base language, but I do not think they are backwards compatible. In other words you can't start with the syntax rules of Javascript and add more patterns, you need to change some of the base patterns to reflect the changes. That puts them kind of in a no-mans-land as far as language support goes.

See also [this extensive discussion on mapping file types across editors](https://github.com/editorconfig/editorconfig/issues/190) that touches on React dialects (Javascript & Typescript) as examples.

Unfortunately I am not a React developer (of any flavor) and am not really qualified to have an opinion about whether to have unique names (i.e. `reactjs`), derivative names (i.e. `javascriptreact`), or flavor names (i.e. javascript.jsx`). As a vim Plugin maintainer however we do need some sort of standard for this. It comes up over and over again and there are several cases out in the wild that re-invent the wheel because vim doesn't have a native filetype built in.

**Comment by ecklf** [comment]
**Disclaimer**: I am using React for roughly 1,5 years so I am not an expert about the in-depths.

From my understanding JSX extends JavaScript with enabling HTML-like syntax to be interpreted as an object. Only within that markup you can use JSX specific syntax (curly braces) to insert JavaScript expressions. 

```jsx
function formatName(user) {
  return user.firstName + ' ' + user.lastName;
}

const user = {
  firstName: 'Harper',
  lastName: 'Perez'
};

const element = (
  <h1>
    {/* this is a comment */}
    Hello, {formatName(user)}!
  </h1>
);
```
(Source: [React Docs)](https://reactjs.org/docs/introducing-jsx.html)

Filetypes: I would prefer to use ```javascriptreact``` and ```typescriptreact```  because of @brammool's comment and to have a common standard with other popular editors like VSCode.

**Comment by brammool** [comment]
> **Disclaimer**: I am using React for roughly 1,5 years so I am not an expert about the in-depths.
> 
> >From my understanding JSX extends JavaScript with enabling HTML-like syntax to be interpreted as an object. Only within that markup you can use JSX specific syntax (curly braces) to insert JavaScript expressions. 
> 
> ```jsx
> function formatName(user) {
>   return user.firstName + ' ' + user.lastName;
> }
> 
> const user = {
>   firstName: 'Harper',
>   lastName: 'Perez'
> };
> 
> const element = (
>   <h1>
>     {/* this is a comment */}
>     Hello, {formatName(user)}!
>   </h1>
> );
> ```
> (Source: [React Docs)](https://reactjs.org/docs/introducing-jsx.html)
> 
> Filetypes: I would prefer to use ```javascriptreact``` and
> ```typescriptreact```  because of @brammool's comment and to have a
> common standard with other popular editors like VSCode.

OK, let's do this then:

" JavaScript with React
au BufNewFile,BufRead *.jsx			setf javascriptreact


" TypeScript with React
au BufNewFile,BufRead *.tsx			setf typescriptreact


Previously .jsx was used for JavaScript.  To reduce the problems with
this change I'll create placeholder javascriptreact.vim files that
source the javascript.vim files for now.


-- 
hundred-and-one symptoms of being an internet addict:
123. You ask the car dealer to install an extra cigarette lighter
     on your new car to power your notebook.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by wincent** [comment]
> The dot notation for a filetype/syntax name is rarely used and most
> users won't know how it works.  I would prefer using a normal name,
> unless the name with a dot has a clear advantage.

Can we please reconsider this change? I am not sure that it was correct, and based on the number of backlinks in [the GitHub issue](https://github.com/vim/vim/issues/4830) I get the impression that it has caused a lot of churn in the plug-in and dot-file ecosystem, and will likely cause more as more update to newer versions of Vim and NeoVim.

The problem with this change is that the JSX syntax *is* strictly additive, just like Doxygen was in your example, Bram. In both JS and TS, you can only use JSX if you explicitly enable it in your tooling in some way (by passing [the `--jsx` flag](https://www.typescriptlang.org/docs/handbook/compiler-options.html) to the TypeScript compiler in the case of TypeScript, or by using and configurating a [JSX-specific plugin](https://babeljs.io/docs/en/babel-plugin-syntax-jsx) in a transpiler like [Babel](https://babeljs.io/)). As an additional nuance, in JS you could generally use JSX with any number of file extensions (".js", ".jsx", and even things like ".react.js"), but in TS the compiler obliges you to use ".tsx".

In short, whether or not a particular file uses (or may use) JSX is a very context-sensitive thing, and I believe it is not uncommon for people to have some pretty dynamic filetype setting going on that will add it using a dot-based "compound" filetype like "javascript.jsx" based on, for example, whether a ".js" contains the word "React" or a call to `React.createClass` etc in the first N lines.

The dot-based "compound" filetypes had the nice property that they could be combined conveniently in various ways; for example:

- `javascript` = just a JS file
- `javascript.jsx` = a JS file with JSX in it
- `javascript.jest` = a JS file with [Jest](https://jestjs.io) (a test framework) tests in it (for example, a user might want to activate test-specific snippets in such a file).
- `javascript.jsx.jest` = a JS file with JSX and Jest tests in it etc.

Basically, there is a pretty rich ecosystem of additive layers on top of JS and TS, and being able to flexibly combine them via dot-based compound filetypes is incredibly useful. Modeling all of these (and more) possible combinations with uncomposable filetypes would lead to a large number of concrete filetypes that don't really seem called for for these additive use cases. In order to preserve this functionality I'm going to work around it by effectively overriding the "javascriptreact" filetype and setting it back to "javascript.jsx", but I wanted to post here just to flag what looks like might have been a mistake.

One final point: there is nothing specific to React about JSX, so it is unfortunate that the filetype has now intrinsically baked in the word "react". Even though JSX was [invented in the context of React's development](https://reactjs.org/docs/introducing-jsx.html), it is actually just a generic syntax which Facebook has published independently as [a spec](https://facebook.github.io/jsx/). When Google [introduced incremental DOM](https://medium.com/google-developers/introducing-incremental-dom-e98f79ce2c5f), for example, they said:

> We would love to see a JSX implementation based on it.

And sure enough, it [was added](https://github.com/google/incremental-dom/pull/76) to the [incremental-dom package](https://www.npmjs.com/package/incremental-dom), there is a [metal-jsx project](https://metaljs.com/docs/guides/jsx-components.html) that uses JSX, and there is at least one [jsx-transform project](https://github.com/alexmingoia/jsx-transform) that is decoupled from React.

Do you think we can revisit this decision?

**Comment by brammool** [comment]
> > The dot notation for a filetype/syntax name is rarely used and most
> > users won't know how it works.  I would prefer using a normal name,
> > unless the name with a dot has a clear advantage.
> 
> Can we please reconsider this change? I am not sure that it was
> correct, and based on the number of backlinks in [the GitHub
> issue](https://github.com/vim/vim/issues/4830) I get the impression
> that it has caused a lot of churn in the plug-in and dot-file
> ecosystem, and will likely cause more as more update to newer versions
> of Vim and NeoVim.
> 
> The problem with this change is that the JSX syntax *is* strictly
> additive, just like Doxygen was in your example, Bram. In both JS and
> TS, you can only use JSX if you explicitly enable it in your tooling
> in some way (by passing [the `--jsx`
> flag](https://www.typescriptlang.org/docs/handbook/compiler-options.html)
> to the TypeScript compiler in the case of TypeScript, or by using and
> configurating a [JSX-specific
> plugin](https://babeljs.io/docs/en/babel-plugin-syntax-jsx) in a
> transpiler like [Babel](https://babeljs.io/)). As an additional
> nuance, in JS you could generally use JSX with any number of file
> extensions (".js", ".jsx", and even things like ".react.js"), but in
> TS the compiler obliges you to use ".tsx".

I thought the original explanation was about React.  What is the full
name for "jsx" then?  The "js" suggests JavaScript, but it can also be
used with Typescript.

If JSX is additive, and the syntax does not depend on the language it's
used on top of, then we could indeed use the dot form.

> In short, whether or not a particular file uses (or may use) JSX is a
> very context-sensitive thing, and I believe it is not uncommon for
> people to have some pretty dynamic filetype setting going on that will
> add it using a dot-based "compound" filetype like "javascript.jsx"
> based on, for example, whether a ".js" contains the word "React" or a
> call to `React.createClass` etc in the first N lines.
> 
> The dot-based "compound" filetypes had the nice property that they could be combined conveniently in various ways; for example:
> 
> - `javascript` = just a JS file
> - `javascript.jsx` = a JS file with JSX in it
> - `javascript.jest` = a JS file with [Jest](https://jestjs.io) (a test framework) tests in it (for example, a user might want to activate test-specific snippets in such a file).
> - `javascript.jsx.jest` = a JS file with JSX and Jest tests in it etc.
> 
> Basically, there is a pretty rich ecosystem of additive layers on top
> of JS and TS, and being able to flexibly combine them via dot-based
> compound filetypes is incredibly useful. Modeling all of these (and
> more) possible combinations with uncomposable filetypes would lead to
> a large number of concrete filetypes that don't really seem called for
> for these additive use cases. In order to preserve this functionality
> I'm going to work around it by effectively overriding the
> "javascriptreact" filetype and setting it back to "javascript.jsx",
> but I wanted to post here just to flag what looks like might have been
> a mistake.

Thus the JSX syntax has no group names from Javascript or Typescript?

> One final point: there is nothing specific to React about JSX, so it
> is unfortunate that the filetype has now intrinsically baked in the
> word "react". Even though JSX was [invented in the context of React's
> development](https://reactjs.org/docs/introducing-jsx.html), it is
> actually just a generic syntax which Facebook has published
> independently as [a spec](https://facebook.github.io/jsx/). When
> Google [introduced incremental
> DOM](https://medium.com/google-developers/introducing-incremental-dom-e98f79ce2c5f),
> for example, they said:
> 
> > We would love to see a JSX implementation based on it.
> 
> And sure enough, it [was
> added](https://github.com/google/incremental-dom/pull/76) to the
> [incremental-dom
> package](https://www.npmjs.com/package/incremental-dom), there is a
> [metal-jsx
> project](https://metaljs.com/docs/guides/jsx-components.html) that
> uses JSX, and there is at least one [jsx-transform
> project](https://github.com/alexmingoia/jsx-transform) that is
> decoupled from React.
> 
> Do you think we can revisit this decision?

I suppose so.  Does anybody else have an opinion?

-- 
While it's true that many normal people whould prefer not to _date_ an
engineer, most normal people harbor an intense desire to _mate_ with them,
thus producing engineerlike children who will have high-paying jobs long
before losing their virginity.
				(Scott Adams - The Dilbert principle)

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by ecklf** [comment]
For starters, sorry if this caused any trouble. I am happy to join the discussion to find the best solution for a standard filetype.

> In both JS and TS, you can only use JSX if you explicitly enable it in your tooling in some way (by passing [the `--jsx` flag](https://www.typescriptlang.org/docs/handbook/compiler-options.html) to the TypeScript compiler in the case of TypeScript, or by using and configurating a [JSX-specific plugin](https://babeljs.io/docs/en/babel-plugin-syntax-jsx) in a transpiler like [Babel](https://babeljs.io/)). 

>One final point: there is nothing specific to React about JSX, so it is unfortunate that the filetype has now intrinsically baked in the word "react".

As far as I know you would still need to pass a  `--jsx "react"` flag with a custom `--jsxFactory` for that. You are totally right with your last point, but I guess Microsoft thinks otherwise.

> * `javascript.jest` = a JS file with [Jest](https://jestjs.io) (a test framework) tests in it (for example, a user might want to activate test-specific snippets in such a file).
> * `javascript.jsx.jest` = a JS file with JSX and Jest tests in it etc.

Personally I've never seen the `.jest` extension. For example I do know  `file.spec.js` and `file.test.js`, in which test specific snippets would trigger for the filetype `javascript`.

> Modelling all of these (and more) possible combinations with uncomposable filetypes would lead to a large number of concrete filetypes that don't really seem called for for these additive use cases.

What if I want concrete snippets for React? If all *flavoured* snippets will trigger for the filetype `javascript.jsx` / `typescript.tsx` it will feel quite messy.

I am aware you described a potential solution:

> In short, whether or not a particular file uses (or may use) JSX is a very context-sensitive thing, and I believe it is not uncommon for people to have some pretty dynamic filetype setting going on that will add it using a dot-based "compound" filetype like "javascript.jsx" based on, for example, whether a ".js" contains the word "React" or a call to `React.createClass` etc in the first N lines.

The question is whether to go through that effort or adapt to potential changes (I.e a big new Framework that works with JSX). Is there any data on how many people use JSX outside of React?

> Do you think we can revisit this decision?

I think you are right. It feels weird to have `.jsx` / `.tsx` bound to the name `react`. However it is undeniable that VSCode, TypeScript and thus Microsofts "own" standards seem to be quite popular. 

Don't get me wrong, I will be happy with either decision. Generally I would still tend to work together on a common standard.

**Comment by clentfort** [comment]
There are a number of frameworks that support JSX to just name a few from the top of my head:

- Vue https://vuejs.org/v2/guide/render-function.html
- Preact https://preactjs.com
- Mithril https://mithril.js.org/jsx.html

But there are more ways that JSX is used in the JavaScript/TypeScript community. As @wincent mentioned you can configure TypeScript as well as Babel to use other functions than `React.createElement` to translate JSX-element creation.

**Comment by ecklf** [comment]
> There are a number of frameworks that support JSX to just name a few from the top of my head:
> 
> * Vue https://vuejs.org/v2/guide/render-function.html
> * Preact https://preactjs.com
> * Mithril https://mithril.js.org/jsx.html
> 
> But there are more ways that JSX is used in the JavaScript/TypeScript community. As @wincent mentioned you can configure TypeScript as well as Babel to use other functions than `React.createElement` to translate JSX-element creation.

Just did some googling and couldn't find an example where those Frameworks actually use the `.jsx` suffix in tutorials / boilerplates. Seems even the React Maintainers do not encourage using it anymore: 
https://github.com/facebook/create-react-app/issues/87#issuecomment-234627904.

However the readme of [coc-tsserver](https://github.com/neoclide/coc-tsserver) says the following:

>Note: for React to work as expected, you need your JSX filetype to be javascript.jsx and your TSX filetype to be typescript.jsx or typescript.tsx. In coc.nvim, these filetypes are mapped to javascriptreact and typescriptreact because that's what tsserver uses.

It does not mention anything but React. I could reach out to the maintainer and ask him how this affects other Frameworks? In that case the `.jsx` suffix would only be a necessity for React and could keep a concrete name.

**Comment by alerque** [comment]
It seems to me there is some conflation here between languages and things done with those languages. Perhaps it's my ignorance of of the JS ecosystem talking here, but Javascript + Jest doesn't sound like a different file type to me, it sounds like a JS file that happens to rely on a specific test framework -- of which there are hundreds. But we don't use file types in general te specify what's being used in those files, just the language they are written in. For example the lua files I'm working on right now are `ft=lua` even though most of the content uses the Busted test framework. I've had to tell my linter that these files are for Busted so it knows what variable scopes are allowed, but I don't need to change the filetype for that.

React is a bit of a different story because it fundamentally changes the syntax of the language, not just embedding another language but merging different syntaxes.

How many of those are there? Does Jest change the language itself? Vue? Foo? Is this an N+1 problem for based languages, or are there really just a bunch of different frameworks that are all js+jsx under the hood?

**Comment by chemzqm** [comment]
React syntax is addtion of javascript,  we've been using javasctipt.jsx for some years, we don't need javascriptreact unless we need talk to tsserver or extension of vscode.

**Comment by wincent** [comment]
@impulse

> Just did some googling and couldn't find an example where those Frameworks actually use the `.jsx` suffix in tutorials / boilerplates.

Please don't confuse the *file extension* with the *file type* (Vim `'filetype'`). I agree with you that it's less common now to see a ".jsx" *extension* (although I'm sure you could if you googled just a *little* bit harder), but that's irrelevant to this discussion, which is about *file types*.

>  you need your JSX filetype to be javascript.jsx and your TSX filetype to be typescript.jsx or typescript.tsx. In coc.nvim, these filetypes are mapped to javascriptreact and typescriptreact because that's what tsserver uses.

Vim has a rich filetype mechanism that allows you to create "compound" filetypes by joining together several types with a dot. The LSP has no such notion, which is why the coc.nvim author maps from the Vim filetype to one that the LSP will understand. But that is not an argument for dumbing down the Vim filetype handling to the lowest common denominator, which is the simple LSP types.

@alerque 

> Perhaps it's my ignorance of of the JS ecosystem talking here, but Javascript + Jest doesn't sound like a different file type to me

I gave an example above of why `javascript.jest` is a useful filetype (the activation of framework-specific snippets). It is definitely not a different language, but neither is a C file a different dialect of C just because it has Doxygen comments in it: the point is that Vim has a flexible and sophisticated filetype system that allows you to model subtleties (in this case additive features layered on top of a base) that go beyond the mere identity of the language.

**Comment by ecklf** [comment]
@wincent 

> Please don't confuse the file extension with the file type (Vim 'filetype'). I agree with you that it's less common now to see a ".jsx" extension (although I'm sure you could if you googled just a little bit harder), but that's irrelevant to this discussion, which is about file types.

I think you misunderstood me. The point I was trying to make is that you can still use JSX within a `.js` file that will have the filetype `javascript`. Even with syntax highlighting https://github.com/facebook/create-react-app/issues/87#issuecomment-239079978

My thought was if filetype `javascript.jsx` / `javascriptreact` (set based on suffix `.jsx`)  is only relevant for React anyways, why not just use the filetype `javascript` for everything else and keep the "dumbed down" filetype for the ts-server to function.

**Comment by wincent** [comment]
> My thought was if filetype `javascript.jsx` / `javascriptreact` (set based on suffix `.jsx`) is only relevant for React anyways

But it's not "only relevant for React"; see the multiple links above to non-React usages of JSX.

> why not just use the filetype javascript for everything else and keep the "dumbed down" filetype for the ts-server to function.

Because it doesn't seem to provide any value to do so; it's effectively a breaking change (see the pingbacks on this issue), and as @chemzqm said (emphasis added):

> React syntax is addtion of javascript, **we've been using javasctipt.jsx for some years, we don't need javascriptreact**.

The right place to do the mapping from the world-of-Vim `'filetype'` to the world-of-LSP language name is inside the plug-in, which is exactly where it already was.

**Comment by ecklf** [comment]
> But it's not "only relevant for React"; see the multiple links above to non-React usages of JSX.

Right, but you are not forced to use `.jsx` to actually support JSX. The only exception is TypeScript with Microsofts standards.

> Because it doesn't seem to provide any value to do so; it's effectively a breaking change (see the pingbacks on this issue), and as @chemzqm said (emphasis added):

For me there kind of is. Whether you set filetype `javascript.jsx` or anything more declarative in vim, for the LSP you will still be in "React world". If there will be more framework specific filetypes with tailored extensions and snippets, you will need to build in a detection system just to remap to a filetype the LSP will understand.

> If JSX is additive, and the syntax does not depend on the language it's
used on top of, then we could indeed use the dot form.

I feel I am still quite hindered by the fact that JSX (as mentioned in the linked [specs](https://facebook.github.io/jsx/)) is tied to one specific language (ECMAScript) while something like `.doxygen` works with multiple.

**Comment by wincent** [comment]
> Right, but you are not forced to use `.jsx` to actually support JSX. The only exception is TypeScript with Microsofts standards.

Again, you seem to be confusing talking about the ".jsx" *file extension* while this thread is about the Vim `'filetype'`. 

> If there will be more framework specific filetypes with tailored extensions and snippets, you will need to build in a detection system just to remap to a filetype the LSP will understand.

Except, you don't. As @chemzqm said, plug-ins have been doing the mapping for you "for some years" and "we don't need javascriptreact". Perhaps you know this, but in case it isn't obvious to others reading this thread, @chemzqm is the maintainer of coc.nvim and the authority you previously called on to support your argument, but he explicitly does the opposite.

 > I feel I am still quite hindered

What is the hindrance exactly? The change you have made here *is a breaking change* for a bunch of people. The bar for justifying such breakage should be considerably high.

> is tied to one specific language

[TypeScript](https://www.typescriptlang.org) *is not* JavaScript, it is "a typed superset of JavaScript that compiles to plain JavaScript". Don't let the close relationship of the languages deceive you into thinking that they are the same language; if they were the same, then you'd be able to paste a snippet of TypeScript into your browser console...

```typescript
function some<O extends object>(o: O, f: <K extends keyof O>(v: O[K], k: K) => boolean): boolean {
  return keys(o).some(k => f(o[k], k))
}
```

...and have it do something other than spit out "Uncaught SyntaxError: Unexpected token '<'", but of course, you can't, because they are different languages.

**Comment by ecklf** [comment]
> Again, you seem to be confusing talking about the ".jsx" file extension while this thread is about the Vim 'filetype'.

vim filetypes are set based on filename extensions. I don't see a reason why this is an "off topic" thing to bring up.

> Except, you don't. As @chemzqm said, plug-ins have been doing the mapping for you "for some years" and "we don't need javascriptreact". Perhaps you know this, but in case it isn't obvious to others reading this thread, @chemzqm is the maintainer of coc.nvim and the authority you previously called on to support your argument, but he explicitly does the opposite.

I don't really think he does state the opposite, he just stated that it does not really matter what filetype will be the standard. Yes `coc.nvim` does the mapping for you, but those are extra lines which will potentially grow (as stated before) bigger with more concrete LSP filetypes. It is not that big of a deal, just something I wanted to add for consideration.

> The change you have made here is a breaking change for a bunch of people. The bar for justifying such breakage should be considerably high.

I am by no means justifying it. As said earlier I am happy with either.

**Comment by wincent** [comment]
> vim filetypes are set based on filename extensions. I don't see a reason why this is an "off topic" thing to bring up.

Purely extension-based filetyping is only the simple case. In this discussion itself we've already noted that a file extension of ".js" alone doesn't distinguish between a file that is merely JS or one that is JS with JSX syntax inside it. `ftdetect` files typically look inside a file to figure out the difference. Same thing when multiple languages claim the same file extension (eg. ".m" which could be Objective-C, Matlab etc; ".md" which could be Markdown or Modula-2 etc).

**Comment by ecklf** [comment]
> Same thing when multiple languages claim the same file extension (eg. ".m" which could be Objective-C, Matlab etc; ".md" which could be Markdown or Modula-2 etc).

Those are fundamentally different languages though. While I agree that TypeScript is not JavaScript, it is very tied to it (valid JS = valid TS). I don't know how filetypes are handled for additional syntax in a specific language +  its superset(s) in vim. I guess it's time to wait on intel from the author.

**Comment by yuezk** [comment]
Hi, @brammool I'm here to give a kindly remind in case you forgot it. 

I think it might be a critical issue as it has affected a bunch of plugins. The good news is that it has been discussed thoroughly.

So, we are waiting for your decision now.

**Comment by brammool** [comment]
I see a lot of discussion but no clear proposal.  If we have to make a choice between using "javascript.jsx" or "javascriptreact", please summarize the advantages and disadvantages and propose a choice.
At least consider creating javascriptreact indent/syntax files that load the javascript indent/syntax, and then some "jsx" on top, versus that "jsx" not being aware of what file type it's being put on top of.

**Comment by ecklf** [comment]
@brammool my takeaways from this discussion:

`javascript.jsx` / `typescript.tsx`

Pro:
- less (probably none?) breaking changes in current plugin ecosystem
- more descriptive name to indicate a file contains JSX syntax

Contra:
- will require additional plugin code to detect and remap to LSP filetype
- dot syntax would be used for one specific language JavaScript and its superset TypeScript

`javascriptreact` / `typescriptreact`

Pro:
- LSP standard -> no remap needed for language server
- describes specific syntax additions for JavaScript and its superset TypeScript

Contra:
- "react" in name (not the only use case, but the most common)
- reportedly causes breaking changes

>At least consider creating javascriptreact indent/syntax files that load the javascript indent/syntax, and then some "jsx" on top,

I think this is already done in popular jsx / tsx syntax plugins:
https://github.com/HerringtonDarkholme/yats.vim/blob/7ea1a4da5ab0d3ae73ea6af80b7252ae960e126f/indent/typescriptreact.vim

Personally I didn't have much issues with plugins and `javascriptreact`. I am not a plugin author though, so I cannot evaluate how bad of a "breaking" change this actually is. For me your initial example with `.doxygen` is still valid, since that actually affects multiple languages and not one language and its superset.

**Comment by yuezk** [comment]
I personally don't support the new filetypes. We have been using `javascript.jsx` for years and everything works well until this commit was merged.

As a React developer, I never use the `.jsx` as the extension and the React team doesn't recommend the `.jsx` extension as well (https://github.com/facebook/create-react-app/issues/87#issuecomment-234627904).

> The distinction between .js and .jsx files was useful before Babel, but it’s not that useful anymore.
>
> There are other syntax extensions (e.g. Flow). What would you call a JS file that uses Flow? .flow.js? What about JSX file that uses Flow? .flow.jsx? What about some other experimental syntax? .flow.stage-1.jsx?
>
> Most editors are configurable so you can tell them to use a JSX-capable syntax scheme for .js files. Since **JSX (or Flow) are strict supersets of JS**, I don’t see this as an issue.


Look at what the plugin maintainers did to fix the breaks after this commit merged.
They just move the logic for jsx/tsx into the `javascriptreact.vim` or add a new filetype.

As for me, as a [plugin](https://github.com/MaxMEllon/vim-jsx-pretty) maintainer, what I should do is create a new the `typescriptreact.vim` file and add the following code into it and nothing else. How redundant it is.

```vim
source <sfile>:h/typescript.vim
```

> * will require additional plugin code to detect and remap to LSP filetype

@impulse I think we shouldn't have to care about the LSP filetypes. Vim and LSP are totally two different things. If we need both to communicate, we definitely need an adapter layer to adapt it, right? That's where the filetype re-mapping and other adaptor logic happens.

Besides, what if there are some other filetypes in Vim which don't match the filetypes in LSP, should we change them to adapt to LSP as well? Furthermore, do the other editors need to change their filetypes to adapt to the LSP if the filetypes are different?

BTW, I don't understand why LSP needs to distinct `javascript` and `javascriptreact`.

For my own, I don't even think we need the `javascript.jsx` filetype. `javascript` is enough.

@brammool It's clear that @wincent has got the most upvotes. I think the commit for this issue should be reverted at least. Leave it for the plugin developers.

**Comment by ecklf** [comment]
@yuezk 

> Furthermore, do the other editors need to change their filetypes to adapt to the LSP if the filetypes are different?

https://code.visualstudio.com/docs/languages/identifiers
https://github.com/microsoft/TypeScript-Sublime-Plugin#note-about-tmlanguage-related-issues

> As the TypeScript and TypeScriptReact .tmLanguage definition files are shared across multiple editors including Sublime Text, Atom-TypeScript, and Visual Studio Code, we decided to create a dedicated repo for these files to combine the efforts for improvement. The new repo is at https://github.com/Microsoft/TypeScript-TmLanguage, and all future tmLanguage-related issues will be tracked there and ported back to this repo.

> BTW, I don't understand why LSP needs to distinct javascript and javascriptreact.

You _could_ only enable specific snippets for `javascriptreact`. However as mentioned you do not need to use it (and it is not encouraged) so snippets usually are set to also trigger for `javascript`. For TypeScript that's a different story. Also the compiler forces you to use `.tsx` since its type annotation conflicts with JSX syntax. I assume the language server has a more complex logic for type checking when this identifier is used.

**Comment by brammool** [comment]
Can we add the "jsx" syntax to "javascript" when a global variable is set?
E.g. ":let g:javascript_with_jsx = 1".  Then when the file has the .js extension but also has the jsx syntax items, nothing else needs to be done.

I haven't actually seen any jsx syntax, thus I'm still guessing how this woks.
At least setting 'filetype' to 'javascript.jsx' does not give an error when there is no "jsx" syntax.
The typescript syntax mentions "jsxRegion" but it's not used or defined anywhere.

**Comment by yuezk** [comment]
@brammool The jsx syntax extends the javascript syntax by just adding a value expression to it ([spec](https://facebook.github.io/jsx/)). Below is the example:

```jsx
// Using JSX to express UI components.
var dropdown =
  <Dropdown>
    A dropdown list
    <Menu>
      <MenuItem>Do Something</MenuItem>
      <MenuItem>Do Something Fun!</MenuItem>
      <MenuItem>Do Something Else</MenuItem>
    </Menu>
  </Dropdown>;

render(dropdown);
```
The syntax is a bit of weird at the first glance, but it's very expressive.

You can treat the entire expression as a new value in JavaScript, just like the numbers, bools or objects, etc. It can appear at any position where the value expression can appear at. So, we'd better not to detect whether the jsx syntax items exists.

IMO, We don't have to add new filetypes for `.js`/`.ts` files, keep it as `javascript`/`typescript`.

And for the `.jsx`/`.tsx` files, the Vim community has been using the `javascript.jsx`/`typescript.tsx` as filetypes for years, it works well.

So, my proposal for this issue is use `javascript.jsx` and `typescript.tsx` for `.jsx`/`.tsx` files, or revert the commit, let the community to decide it.

**Comment by alerque** [comment]
I've been trying to work out what is at stake here.

First, the biggest complaint seems to be that this change breaks many plugins from the ad-hock "standard" way of hacking this together, the dot notation. While this is something to consider, it also seems like since there was no official support before and most of the time this is a couple-line change to adapt to plugins that were using an add-hoc non standard before shouldn't be too surprised to adjust now that official support is landing. Sure it takes a little adjustment, but once adjusted most of these don't seem to have any  lasting issues with that.

Second, the most common (not the biggest, just frequently cited) argument _for_ the change seems to be it makes equating it to LSP types easier. Much as in point 1 this seems like a minor consideration as most other filetypes need mapping anyway so two more or less makes no difference. Plugins that handle this can adapt either way.

It seems to be like the biggest real issue at stake is what @wincent brought up in regard to "composability" of filetypes vs @brammool's initial argument about whether the filetype was actually additive.

I'm actually attracted by the concept @wincent suggests, but I'm having a hard time seeing how it applies to this situation at all. After adding JSX syntax to a file it is no longer valid Javascript. In that sense JSX files are not even Javascript files at all, they are an extended *variant* of Javascript that generic non-JSX tooling often can't process or even pass-through.

By using "composed" filetypes such as "javascript.jsx" I would expect all plugins that work on Javascript files to also work on jsx files, but this is not the case. Most syntax highlighters, compilers, linters, etc. are going to choke because the base syntax of the language is different.

If all Javascript tooling could work on JSX files then the dot notation would make sense to me, but in this case it seems like only JSX aware Javascript tooling is actually able to handle them properly. This means that having a separate filetype for them makes a lot more sense to me. Given that I don't see any tooling that allows JSX to be added to anything _other_ than ECMA Script bases and no Javascript tooling works properly on these sources without being made specifically aware of it, having a dedicated filetype makes sense to me.

**Comment by yuezk** [comment]
@alerque Thanks for your summary, I have some thoughts on this.

> By using "composed" filetypes such as "javascript.jsx" I would expect all plugins that work on Javascript files to also work on jsx files, but this is not the case. Most syntax highlighters, compilers, linters, etc. are going to choke because the base syntax of the language is different.

The syntax highlighters, compilers linters would break, but I think that the Vim end users or the plugin maintainers should ensure that these tools should work well with this filetype.

Think about the first time ES6 introduced, did the syntax highlighters, compilers, linters support it? We did introduced the new extensions like `.es`, `.mjs`, but we didn't introduce any new filetype, we use `javascript` for them all.

https://github.com/vim/vim/blob/3b991527e8167f25ad1dfe780b9633c153600955/runtime/filetype.vim#L804

In some ways, we can think the jsx sytnax as the new feature of JavaScript, just like the ES6/ESNext introduced, execpt that it was not a standard spec.

**Comment by brammool** [comment]
> I've been trying to work out what is at stake here.
> 
> First, the biggest complaint seems to be that this change breaks many
> plugins from the ad-hock "standard" way of hacking this together, the
> dot notation. While this is something to consider, it also seems like
> since there was no official support before and most of the time this is
> a couple-line change to adapt to plugins that were using an add-hoc non
> standard before shouldn't be too surprised to adjust now that official
> support is landing. Sure it takes a little adjustment, but once adjusted
> most of these don't seem to have any  lasting issues with that.
> 
> Second, the most common (not the biggest, just frequently cited)
> argument _for_ the change seems to be it makes equating it to LSP types
> easier. Much as in point 1 this seems like a minor consideration as most
> other filetypes need mapping anyway so two more or less makes no
> difference. Plugins that handle this can adapt either way.
> 
> It seems to be like the biggest real issue at stake is what @wincent
> brought up in regard to "composability" of filetypes vs @brammool's
> initial argument about whether the filetype was actually additive.
> 
> I'm actually attracted by the concept @wincent suggests, but I'm having
> a hard time applying it to this situation at all. After adding JSX
> syntax to a file it is no longer valid Javascript. In that sense JSX
> files are not even Javascript files at all, they are and extended
> variant of Javascript that normal tooling can't even pass through.

That is an important observation.  Other examples where the dot notation
is used, such as using doxygen, do not work like this.  AFAIK they
highlight extra items in comments and do not change the syntax.
 
> By using "composed" filetypes such as "javascript.jsx" I would expect
> all plugins that work on Javascript files to also work on jsx files, but
> this is not the case. Most syntax highlighters, compilers, linters, etc.
> are going to choke because the base syntax of the language is different.
> 
> If all Javascript tooling could work on JSX files then the dot notation
> would make sense to me, but in this case it seems like only JSX aware
> Javascript tooling is actually able to handle them properly. This means
> that having a separate filetype for them makes a lot more sense to me.
> Given that I don't see any tooling that allows JSX to be added to
> anything _other_ than ECMA Script bases and no Javascript tooling works
> properly on these sources without being made specifically aware of it,
> having a dedicated filetype makes sense to me.

-- 
The 50-50-90 rule: Anytime you have a 50-50 chance of getting
something right, there's a 90% probability you'll get it wrong.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by brammool** [comment]
Getting back to this: Suppose I edit a Javascript file with the extra jsx annotations, and only use the Javascript syntax.  If I read the above correctly, then the highlighting would be messed up, because the Javascript syntax highlighter trips over the extra annotations, and text below it is likely not highlighted correctly.  

If that is the case, then it really is a new filetype.  It's not like you can optionally add some highlighting, e.g. for annotations in the comments, like doxygen does.

About the name: We originally used "react", since that was mentioned first.  But "jsx" gets mentioned more often now.  So should we call the filetypes "javascriptjsx" and "typescriptjsx"?

We can add the indent and syntax files for these filetypes that would be similar to the javascript and typescript ones.  However, I don't have a jsx syntax to add on top.  Is there something that can be included in the distribution?

**Comment by brammool** [comment]
Oh, there actually is a typescript + jsx syntax: https://github.com/HerringtonDarkholme/yats.vim/blob/master/syntax/typescriptreact.vim

**Comment by yuezk** [comment]
@brammool Try [vim-jsx-pretty](https://github.com/MaxMEllon/vim-jsx-pretty), it is a syntax highlight plugin for the jsx annotations, it doesn’t rely on the filetype, so it can highlight the source javascript file with extra jsx annotations.

Like I mentioned before, jsx annotations just like the new syntax introduced with the ES6/ESNext. The new syntax (the arrow function, the object destructuring and the class  syntax, etc.) can also break the javascript syntax if the syntax plugin doesn’t adapt it.

The yats plugin you mentioned used the typescriptreact at the very beginning and then it changed to typescript.tsx filetype (https://github.com/HerringtonDarkholme/yats.vim/pull/70), after the commit for this issue merged, it changed to typescriptreact again.

**Comment by brammool** [comment]
Thanks for pointing to that vim-jsx-pretty stuff.  It also makes clear that we are dealing with a specific filetype.  Not that it depends on  leafgarland/typescript-vim, which doesn't have a license, thus you can't legally use it.

I think it's OK to go with the names javascriptreact and typescriptreact.  I see "react" has the name of the filetype, while "JSX" is the name of the items that are added.  Note that for typescript at least one item is omitted, that is using the <> form for type annotations.

I'll update the typescriptreact syntax file.  I don't have an actual javascriptreact syntax file yet.  Currently it just sources the javascript syntax.  Indent plugins are also still missing.

**Comment by yuezk** [comment]
OK, fine. I respect your decision.

**Comment by brammool** [comment]
s/Not that it/Note that it/
Let's close this and hopefully we get contributions for the missing indent and syntax files.

**Comment by alerque** [comment]
> Like I mentioned before, jsx annotations just like the new syntax introduced with the ES6/ESNext.

JSX is not equivalent to ES6 or anything else that aimed to _be_ the next standard specification for the language supported by all interpreters of that language. In fact the JSX spec very specifically says it is not intended to be like that. It is a very domain specific add-on to a base language(s), and the result is not a language backwards compatible with the base it was derived from.

**Comment by yuezk** [comment]
I have something to say about this.

The JSX spec says:

> JSX is an XML-like syntax extension to ECMAScript without any defined semantics. It's NOT intended to be implemented by engines or browsers. It's NOT a proposal to incorporate JSX into the ECMAScript spec itself. It's intended to be used by various preprocessors (transpilers) to transform these tokens into standard ECMAScript.

This paragraph means that the JSX won't introduce any semantics things, it's just an extension of the ECMAScript. And the JSX team also conveys their attitude about the JSX spec to the developers that they won't want to merge the JSX spec into the ECMAScript specs. But this can't change the fact that the JSX syntax is the extension of the ECMAScript. Like it says in the [syntax section](https://github.com/facebook/jsx#syntax).

>JSX extends the PrimaryExpression in the [ECMAScript 6th Edition (ECMA-262)](https://www.ecma-international.org/ecma-262/8.0/index.html#sec-primary-expression) grammar

This means that the JSX syntax should be treated as a primary expression like the other primary expressions (Array Literal, Object Literal, Function Expression, and the Class Expression, etc.) in ECMAScript.

Class Expression，which was introduced with ES6， is also a primary expression and it's not a backward-compatible feature as well, but we didn't add a new filetype to it.
